Azimuthal Equidistant projection

QuLogic opened the pull request:
This is not totally complete yet, but I'm looking for some feedback. Let me preface this by saying I'm neither an expert in cartography nor the proj4 library. I have implemented the CRS, some tests, and added to the documented list of projections (note: this depends on #505).

![projections-3](https://cloud.githubusercontent.com/assets/302469/4852417/d939cb96-607b-11e4-8b79-4e11442a0fe9.png)

However, I am running into a couple minor (or not?) issues:
 * The tests are from the USGS report 1395, [Map Projections - A Working Manual](http://pubs.usgs.gov/pp/1395/report.pdf), but two of them do not work out exactly correct. I tried [implementing the equations myself](https://gist.github.com/QuLogic/0eb4afbc412212b8315a) to see if it was perhaps a precision issue in the report, but I get the same answer.
 * The coastlines will not always draw, depending on the projection parameters. I used the North pole as the centre for the documentation (this is the same as the UN logo), but other central points do not always work. Sometimes they are distorted, and sometimes there's nothing visible at all.
 * None of the internal plumbing is documented, so I'm not really sure about all the CRS methods (e.g., `CRS.threshold`).

pelson commented on docs/make_projection.py:
```diff
@@ -78,31 +63,27 @@ def projection_rst(projection_cls):
+                width = int(width)
+
+            instance_params = ', '.join('{}={}'.format(k, v)
+                                        for k, v in instance_args.items())
```
Would you mind sorting this?

QuLogic replied:
Sorry for being confusing, but this change is actually part of #505, which you've already merged.

pelson commented on lib/cartopy/crs.py:
```diff
@@ -1450,6 +1450,55 @@ def y_limits(self):
+                        ('lat_0', central_latitude),
+                        ('x_0', false_easting), ('y_0', false_northing)]
+        super(AzimuthalEquidistant, self).__init__(proj4_params, globe=globe)
+        self._max = 2e7
```
I suspect this and globe don't play well (globe could be a sphere of radius 1, which would reduce the max). Maybe remove the globe argument for now...

pelson replied:
I may be completely wrong about this - looking at the Orthographic projection suggests that it may be fine.

Could you visually test a highly eccentric spheroid?

QuLogic replied:
I tried the following:
```
import matplotlib.pyplot as plt
import cartopy.crs as ccrs

globe = ccrs.Globe(semimajor_axis=1000, semiminor_axis=500,ellipse=None)
aeqd = ccrs.AzimuthalEquidistant(globe=globe)

plt.figure()
ax = plt.axes(projection=aeqd)
ax.coastlines(resolution='110m')
ax.gridlines()
plt.show()
```
It takes a while, but eventually fails with:
```
Traceback (most recent call last):
  File ".../matplotlib/backends/backend_qt4.py", line 299, in resizeEvent
    self.draw()
  File ".../matplotlib/backends/backend_qt4agg.py", line 154, in draw
    FigureCanvasAgg.draw(self)
  File ".../matplotlib/backends/backend_agg.py", line 451, in draw
    self.figure.draw(self.renderer)
  File ".../matplotlib/artist.py", line 55, in draw_wrapper
    draw(artist, renderer, *args, **kwargs)
  File ".../matplotlib/figure.py", line 1034, in draw
    func(*args)
  File ".../matplotlib/artist.py", line 55, in draw_wrapper
    draw(artist, renderer, *args, **kwargs)
  File ".../cartopy/mpl/geoaxes.py", line 358, in draw
    inframe=inframe)
  File ".../matplotlib/artist.py", line 55, in draw_wrapper
    draw(artist, renderer, *args, **kwargs)
  File ".../matplotlib/axes.py", line 2086, in draw
    a.draw(renderer)
  File ".../matplotlib/artist.py", line 55, in draw_wrapper
    draw(artist, renderer, *args, **kwargs)
  File ".../cartopy/mpl/feature_artist.py", line 103, in draw
    extent = ax.get_extent(feature_crs)
  File ".../cartopy/mpl/geoaxes.py", line 500, in get_extent
    p = self._get_extent_geom(crs)
  File ".../cartopy/mpl/geoaxes.py", line 546, in _get_extent_geom
    self.projection)
  File ".../cartopy/crs.py", line 172, in project_geometry
    return getattr(self, method_name)(geometry, src_crs)
  File ".../cartopy/crs.py", line 306, in _project_polygon
    rings.extend(self._attach_lines_to_boundary(multi_lines, is_ccw))
  File ".../cartopy/crs.py", line 425, in _attach_lines_to_boundary
    raise RuntimeError('Unidentified problem with '
RuntimeError: Unidentified problem with geometry, linestring being re-added
```
and the plot is empty. That being said, I _also_ tried this out with `Orthographic`, and the result is a _very_ zoomed out figure, so I guess there's a bug there too. I will investigate if I can determine the boundary in a bit more generic manner.

pelson replied:
This may be fixed on master now - I've made the geometry projection a little more robust.

pelson commented:
I think this is pretty much good to go. Any reason to hold off of merging @QuLogic?

QuLogic commented:
Well, there is the problem I mentioned above about two tests failing:
```
======================================================================
FAIL: test_ellipsoid_guam_transform (cartopy.tests.crs.test_azimuthal_equidistant.TestAzimuthalEquidistant)
----------------------------------------------------------------------
Traceback (most recent call last):
  File ".../cartopy/tests/crs/test_azimuthal_equidistant.py", line 110, in test_ellipsoid_guam_transform
    assert_array_almost_equal(result, [37712.48, 35242.00], decimal=2)
  File ".../numpy/testing/utils.py", line 842, in assert_array_almost_equal
    precision=decimal)
  File ".../numpy/testing/utils.py", line 665, in assert_array_compare
    raise AssertionError(msg)
AssertionError: 
Arrays are not almost equal to 2 decimals

(mismatch 50.0%)
 x: array([ 37712.47,  35242.02])
 y: array([ 37712.48,  35242.  ])

======================================================================
FAIL: test_ellipsoid_polar_transform (cartopy.tests.crs.test_azimuthal_equidistant.TestAzimuthalEquidistant)
----------------------------------------------------------------------
Traceback (most recent call last):
  File ".../cartopy/tests/crs/test_azimuthal_equidistant.py", line 87, in test_ellipsoid_polar_transform
    assert_array_almost_equal(result, [1078828.3, 289071.2], decimal=1)
  File ".../numpy/testing/utils.py", line 842, in assert_array_almost_equal
    precision=decimal)
  File ".../numpy/testing/utils.py", line 665, in assert_array_compare
    raise AssertionError(msg)
AssertionError: 
Arrays are not almost equal to 1 decimals

(mismatch 100.0%)
 x: array([ -3.0e-09,   9.4e+06])
 y: array([ 1078828.3,   289071.2])

----------------------------------------------------------------------
```

I'm just not sure where to look to attempt to fix them.

pelson commented:
> Well, there is the problem I mentioned above about two tests failing
> I'm just not sure where to look to attempt to fix them. Perhaps there is something obvious I have overlooked.

Are your lats and lons WGS84? You've created a Geodetic with the ellipsoid from the projection, which is emphatically not WGS84.

QuLogic commented:
AFAIUI, the manual uses Clarke 1866 for the Guam example and what it calls the "International ellipsoid" for the polar example. Neither of these appear to be WGS84.

QuLogic commented:
Success! I always thought it was something simple, and it turns out for the ellipsoid transform, I had just accidentally swapped longitude and latitude! For the Guam transform, I had a closer look at the manual, and it turns out they use an approximation for that region. So I reduced the precision for that test, since we aren't using that approximation here.

I haven't really had a chance to change the boundary, but at the least, it's no worse off than the Orthographic projection.

QuLogic commented:
(Test failures are only due to the new PEP8.)

pelson commented:
This doesn't merge cleanly, so would you mind rebasing. That will pick up the PEP8 fixes, also, would you mind adding a comment to the boundary calculation that this doesn't currently account for all of the globes that can be provided.

Other than that :+1: - I'll get this in before cutting v0.12.0rc1. There will be a little bit of work to put together a what's new, and it would be cool if there were a visualisation of this new projection - can you provide some code which demonstrates the projection well? (it doesn't have to do anything other than show the projection)

QuLogic commented:
OK, I'll get to that later today. The projection is plotted in the docs and in the first post here.

pelson commented:
> The projection is plotted in the docs and in the first post here.

A good number of hours later - I have a pretty swish example for the gallery. See #553 :wink:.

QuLogic commented:
> A good number of hours later - I have a pretty swish example for the gallery. See #553 :wink:.

Ohh, I didn't realize you meant a _spiffy_ example. :wink:

pelson commented:
> Ohh, I didn't realize you meant a spiffy example. :wink:

I didn't - I got carried away! I love doing that kind of thing - it gives me an opportunity to be a user - that is an incredibly valuable experience as you remember where the pinch points are :smile:

QuLogic commented:
I rebased and squashed down to the three commits.

Also, I've been [experimenting a bit with the boundaries](http://nbviewer.ipython.org/github/QuLogic/map-projections/blob/master/Azimuthal%20Equidistant.ipynb#Cartopy-Boundaries). After a bit of playing, I realized what probably should have been obvious. Since this plot is azimuthal, the boundary is half a circumference away (e.g., π * 6371 km ~  2e7 m). I guess we'd just need to do something of a similar manner for an ellipsoid.

pelson commented:
Thanks for this @QuLogic - I think this is a really nice addition. I'd love to update the projection to handle crazy ellipsoids, but I didn't want it to hold up this PR.

:+1:
